src/xb-builder-source-ctx.c

```diff
@@ -0,0 +1,165 @@
+	GInputStream		*istream;
+	gchar			*filename;
+	gchar			*content_type;
+} XbBuilderSourceCtxPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (XbBuilderSourceCtx, xb_builder_source_ctx, G_TYPE_OBJECT)
+#define GET_PRIVATE(o) (xb_builder_source_ctx_get_instance_private (o))
+
+/**
+ * xb_builder_source_ctx_get_stream:
+ * @self: a #XbBuilderSourceCtx
+ *
+ * Returns the input stream currently being processed.
+ *
+ * Returns: (transfer none): a #GInputStream
+ *
+ * Since: 0.1.7
+ **/
+GInputStream *
+xb_builder_source_ctx_get_stream (XbBuilderSourceCtx *self)
+{
+	XbBuilderSourceCtxPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (XB_IS_BUILDER_SOURCE_CTX (self), NULL);
+	return priv->istream;
+}
+
+/**
+ * xb_builder_source_ctx_get_bytes:
+ * @self: a #XbBuilderSourceCtx
+ * @cancellable: a #GCancellable, or %NULL
+ * @error: the #GError, or %NULL
+ *
+ * Returns the data currently being processed.
+ *
+ * Returns: (transfer none): a #GInputStream
+ *
+ * Since: 0.1.7
+ **/
+GBytes *
+xb_builder_source_ctx_get_bytes (XbBuilderSourceCtx *self,
+				 GCancellable *cancellable,
+				 GError **error)
+{
+	XbBuilderSourceCtxPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (XB_IS_BUILDER_SOURCE_CTX (self), NULL);
+	return g_input_stream_read_bytes (priv->istream,
+					  128 * 1024 * 1024, /* 128Mb */
+					  cancellable, error);
+}
+
+/**
+ * xb_builder_source_ctx_get_filename:
+ * @self: a #XbBuilderSourceCtx
+ *
+ * Returns the basename of the file currently being processed.
+ *
+ * Returns: a filename, or %NULL if unset
+ *
+ * Since: 0.1.7
+ **/
+const gchar *
+xb_builder_source_ctx_get_filename (XbBuilderSourceCtx *self)
+{
+	XbBuilderSourceCtxPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (XB_IS_BUILDER_SOURCE_CTX (self), NULL);
+	return priv->filename;
+}
+
+/**
+ * xb_builder_source_ctx_get_content_type:
+ * @self: a #XbBuilderSourceCtx
+ *
+ * Returns the content type of the input stream currently being
+ * processed.
+ *
+ * Returns: (transfer full): a content type (e.g. `application/x-desktop`), or %NULL
+ *
+ * Since: 0.1.7
+ **/
+gchar *
+xb_builder_source_ctx_get_content_type (XbBuilderSourceCtx *self,
+					GCancellable *cancellable,
+					GError **error)
+{
+	XbBuilderSourceCtxPrivate *priv = GET_PRIVATE (self);
+	gsize bufsz = 0;
+	guchar buf[4096] = { 0x00 };
+	g_return_val_if_fail (XB_IS_BUILDER_SOURCE_CTX (self), NULL);
+	if (G_IS_SEEKABLE (priv->istream)) {
+		if (!g_input_stream_read_all (priv->istream, buf, sizeof(buf),
+					      &bufsz, cancellable, error))
+			return NULL;
+		if (!g_seekable_seek (G_SEEKABLE (priv->istream), 0, G_SEEK_SET,
+				      cancellable, error))
+			return NULL;
+	}
+	return g_content_type_guess (priv->filename, buf, bufsz, NULL);
+}
+
+/* private */
+void
+xb_builder_source_ctx_set_filename (XbBuilderSourceCtx *self, const gchar *filename)
+{
+	XbBuilderSourceCtxPrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (XB_IS_BUILDER_SOURCE_CTX (self));
```

Comment:
I think you'll want to also fail if filename is NULL too right?

Reply:
Agree, thanks.